share/availability/light/availability_test.go

```diff
@@ -399,38 +384,111 @@ func TestPrunePartialFailed(t *testing.T) {
+	getter := bitswap.NewGetter(ex, clientBs, 0)
+	getter.Start()
+	defer getter.Stop()
+
+	// Create a new ShareAvailability instance and sample the shares
+	sampleAmount := uint(20)
+	avail := NewShareAvailability(getter, ds, clientBs, WithSampleAmount(sampleAmount))
+
+	ctx2, cancel2 := context.WithTimeout(ctx, 1500*time.Millisecond)
+	defer cancel2()
+
+	err := avail.SharesAvailable(ctx2, h)
+	require.Error(t, err, context.Canceled)
+	// close ShareAvailability to force flush of batched writes
+	avail.Close(ctx)
+
+	preDeleteCount := countKeys(ctx, t, clientBs)
+	require.EqualValues(t, sampleAmount, preDeleteCount)
+
+	// prune the samples
+	err = avail.Prune(ctx, h)
+	require.NoError(t, err)
+
+	// Check if samples are deleted
+	postDeleteCount := countKeys(ctx, t, clientBs)
+	require.Zero(t, postDeleteCount)
+
+	// Check if sampling result is deleted
+	exist, err := avail.ds.Has(ctx, datastoreKeyForRoot(h.DAH))
+	require.NoError(t, err)
+	require.False(t, exist)
 }
 
-type fakeSessionExchange struct {
-	exchange.Interface
-	session exchange.Fetcher
+type halfSessionExchange struct {
+	exchange.SessionExchange
+	attempt atomic.Int32
 }
 
-func (fe *fakeSessionExchange) NewSession(context.Context) exchange.Fetcher {
-	return fe.session
+func newHalfSessionExchange(ex exchange.SessionExchange) *halfSessionExchange {
+	return &halfSessionExchange{SessionExchange: ex}
 }
 
-type halfFailBlockstore struct {
-	blockstore.Blockstore
-	attempt atomic.Int32
+func (hse *halfSessionExchange) NewSession(context.Context) exchange.Fetcher {
+	return hse
 }
 
-func newHalfFailBlockstore(bs blockstore.Blockstore) *halfFailBlockstore {
-	return &halfFailBlockstore{Blockstore: bs}
+func (hse *halfSessionExchange) GetBlocks(ctx context.Context, cids []cid.Cid) (<-chan blocks.Block, error) {
+	out := make(chan blocks.Block, len(cids))
+	defer close(out)
+
+	for _, cid := range cids {
+		if hse.attempt.Add(1)%2 == 0 {
+			continue
+		}
+
+		blk, err := hse.SessionExchange.GetBlock(ctx, cid)
+		if err != nil {
+			return nil, err
+		}
+
+		out <- blk
+	}
+
+	return out, nil
+}
+
+type timeoutExchange struct {
+	exchange.SessionExchange
 }
 
-func (hfb *halfFailBlockstore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
-	if hfb.attempt.Add(1)%2 == 0 {
-		return nil, errors.New("fail")
+func newTimeoutExchange(ex exchange.SessionExchange) *timeoutExchange {
+	return &timeoutExchange{SessionExchange: ex}
+}
+
+func (hse *timeoutExchange) NewSession(context.Context) exchange.Fetcher {
+	return hse
+}
+
+func (hse *timeoutExchange) GetBlocks(ctx context.Context, cids []cid.Cid) (<-chan blocks.Block, error) {
+	out := make(chan blocks.Block, len(cids))
+	defer close(out)
+
+	for _, cid := range cids {
+		blk, err := hse.SessionExchange.GetBlock(ctx, cid)
+		if err != nil {
+			return nil, err
+		}
+
+		out <- blk
 	}
-	return hfb.Blockstore.Get(ctx, c)
+
+	// sleep 1 second guarantees that we will exhaust context.
+	time.Sleep(time.Second)
```

Comment:
Why not just wait for context to be finished by reading from <- ctx.Done()?

Reply:
Not that fast. We need to simulate `context.Cancel` error. To do this we need async cancel a context, to be sure that cancelling will happen before return from this method - sleep is needed.

This was the case to simulate `context.DeadlineExceeded` error but have updated to `context.Cancel`. Thanks.